python/ql/test/experimental/dataflow/fieldflow/test.py

```diff
@@ -160,6 +160,162 @@ def test_nested_obj_method():
+    ux(None)
+    SINK_F(foo.x) # $ SPURIOUS: flow="SOURCE, l:-4 -> foo.x"
+
+
+# ------------------------------------------------------------------------------
+# Crosstalk test -- using different function based on conditional
+# ------------------------------------------------------------------------------
+
+class CrosstalkTestX:
+    def __init__(self):
+        self.x = None
+        self.y = None
+
+    def setx(self, value):
+        self.x = value
+
+    def setvalue(self, value):
+        self.x = value
+
+
+class CrosstalkTestY:
+    def __init__(self):
+        self.x = None
+        self.y = None
+
+    def sety(self ,value):
+        self.y = value
+
+    def setvalue(self, value):
+        self.y = value
+
+
+@expects(8) # $ unresolved_call=expects(..) unresolved_call=expects(..)(..)
+def test_no_crosstalk_reference(cond=True):
+    objx = CrosstalkTestX()
+    SINK_F(objx.x)
+    SINK_F(objx.y)
+
+    objy = CrosstalkTestY()
+    SINK_F(objy.x)
+    SINK_F(objy.y)
+
+    if cond:
+        objx.setvalue(SOURCE)
+    else:
+        objy.setvalue(SOURCE)
+
+    SINK(objx.x) # $ flow="SOURCE, l:-4 -> objx.x"
+    SINK_F(objx.y)
+    SINK_F(objy.x)
+    SINK_F(objy.y) # $ flow="SOURCE, l:-5 -> objy.y"
+
+
+@expects(8) # $ unresolved_call=expects(..) unresolved_call=expects(..)(..)
+def test_potential_crosstalk_different_name(cond=True):
+    objx = CrosstalkTestX()
+    SINK_F(objx.x)
+    SINK_F(objx.y)
+
+    objy = CrosstalkTestY()
+    SINK_F(objy.x)
+    SINK_F(objy.y)
+
+    if cond:
+        func = objx.setx
+    else:
+        func = objy.sety
+
+    func(SOURCE)
+
+    SINK(objx.x) # $ MISSING: flow="SOURCE, l:-2 -> objx.x"
+    SINK_F(objx.y)
+    SINK_F(objy.x)
+    SINK_F(objy.y) # $ MISSING: flow="SOURCE, l:-5 -> objy.y"
+
+
+@expects(8) # $ unresolved_call=expects(..) unresolved_call=expects(..)(..)
+def test_potential_crosstalk_same_name(cond=True):
+    objx = CrosstalkTestX()
+    SINK_F(objx.x)
+    SINK_F(objx.y)
+
+    objy = CrosstalkTestY()
+    SINK_F(objy.x)
+    SINK_F(objy.y)
+
+    if cond:
+        func = objx.setvalue
+    else:
+        func = objy.setvalue
+
+    func(SOURCE)
+
+    SINK(objx.x) # $ MISSING: flow="SOURCE, l:-2 -> objx.x"
+    SINK_F(objx.y)
+    SINK_F(objy.x)
+    SINK_F(objy.y) # $ MISSING: flow="SOURCE, l:-5 -> objy.y"
+
+
+@expects(10) # $ unresolved_call=expects(..) unresolved_call=expects(..)(..)
+def test_potential_crosstalk_same_name_object_reference(cond=True):
+    objx = CrosstalkTestX()
+    SINK_F(objx.x)
+    SINK_F(objx.y)
+
+    objy = CrosstalkTestY()
+    SINK_F(objy.x)
+    SINK_F(objy.y)
+
+    if cond:
+        obj = objx
+    else:
+        obj = objy
+
+    obj.setvalue(SOURCE)
+
+    SINK(objx.x) # $ MISSING: flow="SOURCE, l:-2 -> objx.x"
+    SINK_F(objx.y)
+    SINK_F(objy.x)
+    SINK_F(objy.y) # $ MISSING: flow="SOURCE, l:-5 -> objy.y"
+
+    SINK(obj.x) # $ flow="SOURCE, l:-7 -> obj.x"
+    SINK_F(obj.y) # $ flow="SOURCE, l:-8 -> obj.y"
```

Comment:
Oh! No, I think the annotations (by which I mean `SINK` and `SINK_F`) should agree with the runtime. Perhaps we could rewrite to `else if cond1:`?

Reply:
I originally had the following code. I noticed that adding these if-then-else statements made the cfg-splitting kick in, so the problem this is testing was solved trivially. I guess I should have made a comment about this, since it really isn't apparent from the test-case.

```py
if cond:
    SINK(objx.x) # $ MISSING: flow="SOURCE, l:-2 -> objx.x"
    SINK_F(objx.y)
else:
    SINK_F(objy.x)
    SINK(objy.y) # $ MISSING: flow="SOURCE, l:-5 -> objy.y"

if cond:
    SINK(obj.x) # $ flow="SOURCE, l:-7 -> obj.x"
else:
    SINK(obj.y) # $ flow="SOURCE, l:-8 -> obj.y"
```

I agree that `SINK(..., no_present_at_runtime=True)` isn't pretty, but I also agree that my currently solution of using `SINK_F` where `SINK` should have been used is really bad... so I like the `SINK(..., no_present_at_runtime=True)` solution better

Reply:
Ah, but I wanted to rewrite line 293. Would that avoid the problem with extra splitting?

Reply:
I'm confused. I think your line numbers are different than mine?

![image](https://user-images.githubusercontent.com/1054041/203078303-e9e828c8-a06f-4e73-ac42-3a1f7e260f7e.png)

Reply:
Ok, I see the problem, I have found a potential solution, which I will encode in suggestions.

Reply:
Alright, I cannot actually make suggestions in unchanged code, so here is a description of the diff, then you can evaluate if it makes sense :-)

- Change `SINK` to accept an optional expected argument
```python
def SINK(x, expected=None):
    if expected is None and is_source(x):
        print("OK")
    elif expected is not None and x == expected:
        print("OK")
    else:
        print("Unexpected flow", x)
```
This makes `None` illegal as an expected value, though, so
- Change the default values for `CrosstalkTestX` and `CrosstalkTestY` away from `None`
```python
class CrosstalkTestX:
    def __init__(self):
        self.x = NONSOURCE
        self.y = NONSOURCE

    def setx(self, value):
        self.x = value

    def setvalue(self, value):
        self.x = value


class CrosstalkTestY:
    def __init__(self):
        self.x = NONSOURCE
        self.y = NONSOURCE

    def sety(self ,value):
        self.y = value

    def setvalue(self, value):
        self.y = value
```
- Use an inline `if` expression to avoid splitting. Lines 303 and 304 become (hopefully using better line reference numbers)
```python
    SINK(obj.x, SOURCE if cond else NONSOURCE) # $ flow="SOURCE, l:-7 -> obj.x" flow="SOURCE -> IfExp"
    SINK(obj.y, NONSOURCE if cond else SOURCE) # $ flow="SOURCE, l:-8 -> obj.y" flow="SOURCE -> IfExp"
```
This makes both `validTest` and `NormalDataflowTest` happy in my Codespace (although `UnresolvedCalls` and `dataflow-consistency` both have some issues?).

Reply:
I fixed this by adding optional keyword argument `not_present_at_runtime` to the SINK function. I prefer this option from adding `expected` argument, since it's more clear from a reader what is going on.

I fear that an uninitiated person reading `SINK(obj.y, expected=NONSOURCE)` might be confused about whether we expected `obj.y` to contain SOURCE (which is what usually happens with SINK argument) or NONSOURCE (as by the optional expected argument).

I also added comments about the splitting, I realize I should have done this in the beginning :disappointed: 

> Use an inline if expression to avoid splitting. Lines 303 and 304 become (hopefully using better line reference numbers)

in the real call-graph branch, I have a test for such compound arguments, which requires special handling, so this won't work. (I seem to have forgotten to pull it out into this PR :|) -- Added them as a commit now :+1:

Reply:
I ended up using the `not_present_at_runtime` approach, since with `SINK(obj.y, expected=NONSOURCE)` it might be confusing for the non-initiated people whether we expect `obj.y` to be SOURCE (as it usually is for SINK arguments) or NONSOURCE (as indicated by the expected argument).

With `not_present_at_runtime`, it's more clear that the intention is that data-flow finds the flow, but the flow is not present at runtime.

I've also added comments explaining why we can't just use if-then-else -- I should have done this initially, so I'm happy we've had this conversation, so I could add least get those added :sweat: 

> Use an inline if expression to avoid splitting. Lines 303 and 304 become (hopefully using better line reference numbers)

This won't work. Turns out I had forgotten to pull out another test-case from the call-graph branch, so I've also added this as a new commit.

Reply:
I am not sure I agree that overriding the expected value is confusing. What annoys me a bit about `not_present_at_runtime` is that it goes against the spirit of the tests: That the runtime behaviour is our source of truth. I can see that we might at times need to express the approximations of our analysis, but I hope that `not_present_at_runtime` will be used seldomly and only as a last resort..

Reply:
> I hope that not_present_at_runtime will be used seldomly and only as a last resort..

Agreed on that part :+1: